### DISCUSSION TITLE
Deprecate atmosphere packages?

### DISCUSSION THREAD

**Discussion by CaptainN** [meteor/meteor#11765] [Feature Requests]
It's been a while since we discussed this, and I don't remember all the reasons we haven't done this yet. Here's the suggestion, in rough form:

- Deprecate atmoshpere packages entirely.
- Extend npm package.json (for packages) to add Meteor specific items, like the server/client bundle entry points, and non-js static files, npm dependencies, etc.
- Migrate (over time) Meteor core packages to `@meteor/***` - basically, have them be like `@meteor/react-meteor-data`, `@meteor/accounts-base`, and then deprecate the atmosphere packages. Import resolution might depend on the specific way meteor packages are "simulated" from npm packages, if that's the implementation we go with (more below).
- Maybe there's even a way to directly compile for npm using the existing package.js file in many cases.

How?

To make this work, we can either package in the `package.js` file as is in the npm package, and have meteor use that from the node_modules folder (even if it has to translate to an internal package on install/update time - when doing `npm i @meteor/whatever`) - or, we can generate the `package.js` file from the more declarative static extensions to `package.json`. (I like the second option - I very rarely see scripted configuration in `package.js`, and I don't think we need it.)

My main thought about how to leverage the above would be to have Meteor do a "post install" routine when a meteor package is installed from npm (on startup, the way meteor currently reconciles atmosphere deps), to create a stub meteor package from the npm package. Internally, meteor's bundler would still pretty much work the way it does now (minus dependency management), but it would gather the package from node_modules. Packages would still be precompiled, just like they are now - they'd just be distributed through npm.

Why?

Mostly to have only one package system, npm. But also because it simplifies dependency management, by using just npm's package dependency system. It would also grant us new capabilities, such as gaining the ability to use typescript for packages, without having to go through separate @types definitions.

Transition:
This doesn't have to be done all at once. Existing packages could stay in atmosphere, while select packages are ported, and/or new packages are distributed through npm.

Exceptions:
Maybe build packages can stay in atmosphere? IDK enough about them to comment intelligently, except that I do know they are very different from regular packages.

It seems this should be possible, but I don't recall the reasons why we haven't done something like this before.

**Comment by JanMP** [comment]
I just tried writing an npm package. And (as I expected) you can't just use Coffeescript like you can in meteor packages. So my answer to that sugesstion is: HELL NO!!!

**Reply by radekmie** [reply]
Hey, don't worry! You are able to implement them in CoffeeScript and every other language you'd like. The only thing is that you have to _release_ it as JavaScript. There's a lot of packages using TypeScript and yet they work, right?

Of course, that's some additional setup, but it's pretty automatable.

**Reply by CaptainN** [reply]
In reality, Meteor's build tool is building your coffeescript and publishing that as compiled JS. It does that for Typescript and ECMAScript too.

Meteor could still offer tools to make that build process just as easy as it is now, except publish to npm instead of atmosphere. This is not a blocker to migration.

**Reply by JanMP** [reply]
~could~ must

And since we are adding a new build target, maybe we could also add a build target for web-workers while we are at it.

**Comment by radekmie** [comment]
I really like this idea. It would get rid of the need for Meteor packages checking the versions of npm packages. Also, as you've said, it'd be obvious how to provide typings for them (the answer is: as usual).

As for the automatic Atmosphere->npm transformation, I'd go with something different than `@meteor`, e.g., `@meteor-package` - otherwise people may consider these packages as the official npm ones.

I'm not sure how isopacks are structured, but I think it's feasible to consider it in the future.

**Reply by WilliamKelley** [reply]
I agree with publishing under a different scope than `@meteor`, but I'd suggest `@atmosphere`? I think carrying over the name lets users carry over their understanding of what atmosphere is: a meteor-specific package registry (now an alternate-registry scope). But to newcomers it won't be as straightforward as `@meteor-packages`.

**Comment by Saeeed-B** [comment]
I think the atmosphere is obsolete right now.
Users do not trust it enough at all.
Packages are difficult to update and issues are not followed up.
Joining npm requires courage.
